javascript/ql/src/semmle/javascript/security/dataflow/TaintedLengthCustomizations.qll

```diff
@@ -0,0 +1,285 @@
+/**
+ * Provides default sources, sinks and sanitisers for reasoning about
+ * DOS attacks using objects with unbounded length property. 
+ * As well as extension points for adding your own.
+ */
+
+import javascript
+
+module TaintedLength {
+  import semmle.javascript.security.dataflow.RemoteFlowSources
+  import semmle.javascript.security.TaintedObject
+  import DataFlow::PathGraph
+
+  /**
+   * Holds if an exception will be thrown whenever `e` evaluates to `undefined` or `null`.
+   */
+  predicate isCrashingWithNullValues(Expr e) {
+    exists(ExprOrStmt ctx |
+      e = ctx.(PropAccess).getBase()
+      or
+      e = ctx.(InvokeExpr).getCallee()
+      or
+      e = ctx.(AssignExpr).getRhs() and
+      ctx.(AssignExpr).getLhs() instanceof DestructuringPattern
+      or
+      e = ctx.(SpreadElement).getOperand()
+      or
+      e = ctx.(ForOfStmt).getIterationDomain()
+    )
+  }
+
+  /**
+   * A loop that iterates through some array using the `length` property. 
+   * The loop is either of the style `for(..; i < arr.length;...)` or `while(i < arr.length) {..;i++;..}`. 
+   */
+  class ArrayIterationLoop extends Stmt {
+    LocalVariable indexVariable;
+    LoopStmt loop;
+
+    ArrayIterationLoop() {
+      this = loop and 
+      exists(RelationalComparison compare, DataFlow::PropRead lengthRead |
+        compare = loop.getTest() and
+        compare.getLesserOperand() = indexVariable.getAnAccess() and
+        lengthRead.accesses(_, "length") and
+        lengthRead.flowsToExpr(compare.getGreaterOperand())
```

Comment:
Could you move `lengthRead` into a field with a getter predicate so it can be used from `LoopSink`.